### ISSUE TITLE
Smart card PIV devices not accessible on Fedora

### ISSUE THREAD

**Issue by jmpolom** [flathub/com.google.Chrome#209]
Chrome does not pickup attached smart card/PIV devices attached and working via the host. Tested on Fedora 38 with a functioning p11-kit and pcsc daemon. Flatpak might be missing pcsc?

**Comment by refi64** [comment]
We [do give access to the pcsc socket](https://github.com/flathub/com.google.Chrome/blob/master/com.google.Chrome.yaml#L19).

**Comment by jmpolom** [comment]
Yes, I noticed. Apparently that is not sufficient on Fedora. I noticed the flatpak is being built from the deb archive. Does the Debian build of Chrome work with pcsc or p11-kit? I would suspect this aspect or some missing libraries in the flatpak.

**Comment by Erick555** [comment]
At least firefox needs pcsc-lite library to be present: https://github.com/flathub/org.mozilla.firefox.BaseApp/blob/branch/22.08/org.mozilla.firefox.BaseApp.json#L61

BTW: [chromium](https://github.com/flathub/org.chromium.Chromium/blob/master/org.chromium.Chromium.yaml#L8) flatpak has neither pcsc permission or the library

**Comment by jmpolom** [comment]
The flathub flatpak of Firefox also does not seem able to pickup attached smart cards either. pcsc-lite is likely unnecessary to have in the flatpak if the host runs pcscd and the socket is available to the flatpak app.

Generally you need more than just pcsc to make smart cards work in an application on Linux. If the hosts pcscd socket is available to the flatpak app, what is needed is probably opensc and/or p11-kit.

**Comment by Erick555** [comment]
> pcsc-lite is likely unnecessary to have in the flatpak if the host runs pcscd and the socket is available to the flatpak app.

I recommend reading this [first](https://ludovicrousseau.blogspot.com/2022/02/accessing-smart-cards-from-inside.html).

**Comment by jmpolom** [comment]
> [Here's ](https://ludovicrousseau.blogspot.com/2022/02/accessing-smart-cards-from-inside.html)article about how to connect to pcsc daemon from inside flatpak from pscsc creator.
> 
> opensc can be added in flatpak like https://github.com/flathub/org.mozilla.Thunderbird/pull/162 but I couldn't find anyone wanting to test it.

I found and read the referenced article before I posted this issue. Having been using PKCS #11 token smart cards for years on Linux I am acutely aware of the supporting pieces needed to get them working. You need more than pcsc-lite.

opensc is very likely to be the critical missing piece here. I am  unclear if pcsc-lite is needed in the flatpak or not. If the host pcscd socket is available inside the flatpak, the flatpak certainly does not need to run pcscd (host should be relied on to provide low level access to card/reader). On Fedora, I do not see opensc linking to any pcsc libs:

```
$ ldd /usr/lib64/libpcsclite.so.1
	linux-vdso.so.1 (0x00007ffc2a7a3000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f353693e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3536b3a000)
$ ldd /usr/lib64/libopensc.so.8.1.0
	linux-vdso.so.1 (0x00007ffe9c7fd000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f101febe000)
	libcrypto.so.3 => /lib64/libcrypto.so.3 (0x00007f101fa00000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f101f822000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f10200d1000)
$ ldd /usr/lib64/opensc-pkcs11.so
	linux-vdso.so.1 (0x00007ffd79385000)
	libopensc.so.8 => /lib64/libopensc.so.8 (0x00007f0c76e22000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f0c76e08000)
	libcrypto.so.3 => /lib64/libcrypto.so.3 (0x00007f0c76800000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f0c76c2a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f0c77059000)
```

In the above, the opensc-pkcs11.so lib is the one that applications like Thunderbird or Firefox or Chrome will need to load to access PKCS #11 tokens. Fedora ships with p11-kit configured to aggregate and proxy PKCS #11 to applications like Firefox. This obviously does not work inside of flatpaks though that are separated from the rest of the system.

Per the [opensc linux build docs](https://github.com/OpenSC/OpenSC/wiki/Compiling-and-Installing-on-Unix-flavors) pcsc-lite is called out as a build dependency. I am unclear if it is only a build dependency or also a runtime dependency. Have never built opensc from source.

To clarify, pcsc-lite and the pcscd daemon are "lower level" in the stack. They communicate with the card reader. They do not provide any higher level PKCS #11 APIs which is the issue here. PIV uses PKCS #11.

**Comment by Erick555** [comment]
> If the host pcscd socket is available inside the flatpak, the flatpak certainly does not need to run pcscd

That's for sure but who from sandbox do the talking to the socket? From the linked article I got impression it's the client pcsc library.

**Comment by MarceloAlm** [comment]
Hi, I started to use the chrome flatpak version now and am unable to use my USB "Aladdin Knowledge Systems Token JC", Safenet eTOKEN.
To install it, I followed the guide https://wiki.archlinux.org/title/Smartcards item 3.4:
install the proprietary libs (/usr/lib/libeToken.so)
add the config to ~/.pki/nssdb
$ modutil -dbdir sql:$HOME/.pki/nssdb/ -add "eToken" -libfile /usr/lib/libeToken.so

This works for the Chromium browser instaled with the system package, and also when I install chrome directly.
I still trying to understand what is not allowed for the flatpak and stoped at the lack of the library libeToken.so. The permission to the folder ~/.pki/nssdb is easy to solve adding the permission, but the library is not avaliable to the sandbox.

There is a way to include my native library to a flatpak?

**Comment by jmpolom** [comment]
> Hi, I started to use the chrome flatpak version now and am unable to use my USB "Aladdin Knowledge Systems Token JC", Safenet eTOKEN.

This issue is for access to pkcs#11 devices. If it can't be accessed using pcscd and opensc and requires a proprietary so then this issue is likely not relevant.

I posted a [workaround](https://gist.github.com/jmpolom/cbecace9c56f6de72df9c129058caff2) in https://github.com/flatpak/flatpak/issues/5756 for enabling flatpak sandbox access to the p11-kit user socket that you could try.

**Comment by MarceloAlm** [comment]
> > Hi, I started to use the chrome flatpak version now and am unable to use my USB "Aladdin Knowledge Systems Token JC", Safenet eTOKEN.
> 
> This issue is for access to pkcs#11 devices. If it can't be accessed using pcscd and opensc and requires a proprietary so then this issue is likely not relevant.
> 
> I posted a [workaround](https://gist.github.com/jmpolom/cbecace9c56f6de72df9c129058caff2) in [flatpak/flatpak#5756](https://github.com/flatpak/flatpak/issues/5756) for enabling flatpak sandbox access to the p11-kit user socket that you could try.

Thanks for the tip, I will invest some time to understand the pcscd and its components: pkcs11 and pkcs12 (pfx) keys, which I need to access.
You were able to access your certificates with the workaround? I have a question: what is the use of the socket=pcsc permission?

Installed Firefox Flatpak, which is maintained by the same team as the browser, in the hope of reaching a conclusion.
Entered the flatpak with bash command:
$ flatpak run --command=/bin/bash org.mozilla.firefox

The file you mentioned is present in the file system, but I still don't know how to test it.

I think you should take a look at the page https://wiki.archlinux.org/title/Smartcards, perhaps post your solution as a suggestion. Not all Linux users depend on smartcards, but everyone who does will eventually reach this page.

**Comment by MarceloAlm** [comment]
lot of work to solve this, but I managed.

1) install the library for the USB Token (libeToken.so in my case)
2) create a config file for p11-kit to load the module: 
$ sudo echo "module: libeToken.so " > /usr/share/p11-kit/modules/libeToken.module 
or to keep it only for your user
$ mkdir -p ~/.config/p11-kit/modules/
$ echo "module: libeToken.so " > ~/.config/p11-kit/modules/libeToken.module 

The command "p11-kit list-modules" should list the token.

3) the flatpak needs access to the \~/.pki folder and p11-kit:
$ flatpak override --user --filesystem=~/.pki:create com.google.Chrome 
$ flatpak override --user --filesystem=xdg-run/p11-kit/pkcs11 com.google.Chrome

4) copy the libeToken.so to the ~/.pki folder
$ mkdir -p ~/.pki
$ cp /usr/lib/pkcs11/libeToken.so ~/.pki/

4) open the browser to create the NSSDB, then register the library: 
$ flatpak run com.google.Chrome
$ modutil -dbdir sql:$HOME/.pki/nssdb/ -add "eToken" -libfile ~/.pki/libeToken.so

Check the certificates: [chrome://settings/certificates](url)
I hope someone finds a better way to share a custom library to the flatpak.

**Comment by MarceloAlm** [comment]
> The flathub flatpak of Firefox also does not seem able to pickup attached smart cards either. pcsc-lite is likely unnecessary to have in the flatpak if the host runs pcscd and the socket is available to the flatpak app.
> 
> Generally you need more than just pcsc to make smart cards work in an application on Linux. If the hosts pcscd socket is available to the flatpak app, what is needed is probably opensc and/or p11-kit.


Could you try the pull request #378 : https://github.com/flathub/com.google.Chrome/pull/378
With the extension: https://github.com/MarceloAlm/com.google.Chrome.cripto.opensc 
I dont have a opensc compatible token to test it. The eToken extension is working.
